tests/functional/tests/management/test_composite_volume.py

```diff
@@ -216,10 +260,52 @@ def test_io_propagation_basic(pyocf_ctx):
       - composite_volume::volume_api
       - composite_volume::io_request_passing
     """
-    pass
+    pyocf_ctx.register_volume_type(TraceDevice)
+
+    vol_size = S.from_MiB(1)
+    ram_vols = [RamVolume(vol_size * i) for i in range(1, 17)]
+
+    vols, io_count = setup_tracing(ram_vols)
+
+    running_sum = S(0)
+    vol_end = []
+    for v in ram_vols:
+        running_sum += S(v.size)
+        vol_end.append(S(running_sum))
+
+    cvol = CVolume(pyocf_ctx)
+    for vol in vols:
+        cvol.add(vol)
+
+    cvol.open()
+
+    # hit each subvolume at different offset (vol number * 1 KiB)
+    io_addr = [i * S.from_KiB(1) + (vol_end[i - 1] if i > 0 else S(0)) for i, v in enumerate(vols)]
+    io_size = S.from_KiB(12)
+
+    for addr in io_addr:
+        ret = _cvol_io(cvol, addr, io_size, "submit")
+        assert ret == 0
+
+        ret = _cvol_io(cvol, addr, io_size, "submit_flush", IoFlags.FLUSH)
+        assert ret == 0
+
+        ret = _cvol_io(cvol, addr, io_size, "submit_discard")
+        assert ret == 0
+
+    for i, (vol, addr) in enumerate(zip(vols, io_addr)):
+        for io_type in TraceDevice.IoType:
+            ios = io_count[vol][io_type]
+            assert len(ios) == 1
+            io = ios[0]
+            assert io.dir == IoDir.WRITE
+            assert io.addr == addr.B - int(vol_end[i - 1] if i > 0 else 0)
+            assert io.bytes == io_size.B
```

Comment:
it would be more precise to run these checks after each I/O in previous loop

Reply:
changed back to this method